### ISSUE TITLE
RFC: Remove automation from core and move to a new repo

### ISSUE THREAD

**Issue by openhab-5iver** [openhab/openhab-core#1319]
My proposal is to remove all things automation from OHC and move them to a new openHAB repository. I see this new repository being the potential home for the following (maybe more):

* Rule engines
* Scripted automation (if separated out of the NGRE)
* ScriptEngineFactories
* DSL (if kept in OH 3.0)
* Migration tool for .rules files (if the DSL is not kept in OH 3.0)
* Automation bundles
* ModuleTypes
* Rule templates (possibly separated in future)
* ExtensionService for openHAB automation marketplace (or will there be one with a larger scope?)

The NGRE introduces several new pieces to OH automation that should be maintained in a central location. For example, there are Jython (#1253) and JavaScript ScriptEngineFactory PRs pending review, but we first need to decide on where they should be maintained. Right behind these will be Groovy, jRuby, and Kotlin ScriptEngineFactories. There will also soon be a deluge of rule templates. 

I do not think any of these fit well in OHC. Separating automation from OHC will also allow for dependencies on addons to be placed on automation bundles, rule templates, etc., which would not be allowed if these bundles were kept in OHC.

**Comment by kaikreuzer** [comment]
I am not sure I fully understand the proposal.
So far, the structure is that we have "core" components that build the main runtime and that are required for it. The core also defines the APIs for any add-ons etc.
The rule engine and its APIs for potential extensions must imho therefore clearly be a part of the core - or do you consider the ability to execute rules as something optional?

>  there are Jython, JavaScript, Groovy, jRuby, and Kotlin ScriptEngineFactories.

Reading this makes me really worried about how openHAB rule support will look like in future.
With openHAB 3, we took the users feedback and decided to not anymore follow the "open to everything" approach, which heavily confuses people, but to rather limit the available options (e.g. reducing the number of UIs, the number of ways to configure bindings (no 1.x anymore), etc.)

For rules, there should therefore be exactly one way we suggest to write them and this should be used in all documentation and examples - and this also clearly belongs in the core. IF we decide to fully replace DSL by Jython, it would thus mean that Jython scripting needs to be added to the core.

Any "ecosystem" around the rule engine, i.e. any optional stuff for certain use cases and stuff on which no add-on should have any dependency, should not be in OHC, agreed. But for those we should probably define new "add-on categories", clearly document the APIs they are supposed to use and possibly provide skeleton creation scripts. Per default, the best place for such add-ons would be the openhab-addons repo, unless we have very good reasons to decide differently.

Note that any additional repo is a burden for the overall project management (docs, build & release processed, etc.), so you'll have noticed that I am pretty reluctant in this regard ;-)

**Comment by openhab-5iver** [comment]
> The rule engine and its APIs for potential extensions must imho therefore clearly be a part of the core - or do you consider the ability to execute rules as something optional?

Yes, [I view automation as optional](https://github.com/openhab/openhab-core/issues/526#issuecomment-570663344). Officially, we can use the old or new rule engines, but there are also a few 3rd party rule engines available, like Node-RED, HABApp, etc. While the official rule engines are an important part of openHAB and are "core-like", I do not see them as belonging in OHC and feel they should be optional. Same for all of the other bullets in the OP. This is the crux of why I am proposing to separate automation from OHC.

>> there are Jython, JavaScript, Groovy, jRuby, and Kotlin ScriptEngineFactories.

> With openHAB 3, we took the users feedback and decided to not anymore follow the "open to everything" approach, which heavily confuses people, but to rather limit the available options (e.g. reducing the number of UIs, the number of ways to configure bindings (no 1.x anymore), etc.)

I feel very strongly in the importance of an "officially supported" or, at a minimum, "recommended" automation implementation in OH. This will help to focus the development, testing, documentation, and knowledge shared within the forum. There are plenty of options and a lot of them work well, but focusing on just one will minimize effort. 

I don't think this is the case, but one of the things you said was unclear to me and I want to be sure I understood. Are you suggesting to remove the NashornScriptEngineFactory and GenericScriptEngineFactiory so that other script engines cannot be used? I feel that being restrictive is a dead end that will inhibit discovery and invention, and will turn people away from the project. If someone wants to add a KotlinScriptEngineFactory, integration tests, documentation, etc., I do not see this being disruptive to what has been chosen as "officially supported"... it's just another available option. I think a slow organic approach will suffice in preventing the many options from diluting the mainstream, "official", options. But I really do not think this is much of a concern anyhow, due to my next point.

> For rules, there should therefore be exactly one way we suggest to write them and this should be used in all documentation and examples - and this also clearly belongs in the core. IF we decide to fully replace DSL by Jython, it would thus mean that Jython scripting needs to be added to the core.

I think it is best for the future of OH to move past the rules DSL and adopt a pure scripting language. In the future, I envision the large majority of users creating and managing rules through a UI editor, with heavy use of rule templates provided through an ExtensionService. Those rule templates will make use of libraries available in a helper library bundle (currently included in the Jython bundle, but I intend to move them to their own). **In this way, the rules DSL will effectively be replaced more by the UI rule editor than Jython.** However, most of the rule templates will make use of Jython in Scripted Conditions and Actions. Advanced users will not use the UI, but scripted automation with Jython. These are the people who will be contributing the scripts and libraries in rule templates and shared through an ExtensionService. 

My vision of a migration tool has evolved and I now see it building managed JSON rules as a one-time conversion and using Scripted Actions where the rules are too complex to convert to existing Actions. Potentially, the tool could utilize a catalog of rule templates to assist in migrations.

> But for those we should probably define new "add-on categories", clearly document the APIs they are supposed to use and possibly provide skeleton creation scripts. Per default, the best place for such add-ons would be the openhab-addons repo, unless we have very good reasons to decide differently.

So, if someone builds a new rule engine, would it go into addons? If someone builds a full automation solution, like [Area_Triggers_and_Actions](https://openhab-scripters.github.io/openhab-helper-libraries/Python/Community/Area%20Triggers%20and%20Actions.html), would this go into addons? How about a simple automation bundle like Time Of Day? How about a rule template for a motion sensor turning on a light? Maybe it would just take me some time to get used to, but I'm not seeing openhab-addons as a good fit for these, especially if the "official" automation pieces are located in OHC. It makes more sense to me to put _all_ of automation into one repository, whether that is OHC, addons, or a new repo (my preference).

The ExtensionService, as the tool used to distribute shareable bits of automation (other than copy/paste), will have a large impact on the maintenance of automation. I think it would be a mistake to leave everything unchecked/high-risk/unverified/etc. I see two flavors... uncertified and certified (similar to Docker Certified images).

Certified means that the code has been reviewed, not necessarily for functionality, but for safety. I also think we should have "official" rule templates and automation bundles that are reviewed by maintainers, not just to certify them, but to review functionality. Certified does not mean they are "official". Uncertified simply means that they have not gone through certification... they are out in the wild and untested. 

Depending on how the ExtensionService is implemented, I see a LOT of activity with automation bundles, rule templates, and community helper libraries... from users, developers, and maintainers. Another consideration is that the release cycles will be very different for these automation components. I do not see any of this fitting into OHC and is another reason why I am asking for the separation of automation and OHC.

Another big reason to isolate automation from OHC is to provide a focal point, due to it's critical role in openHAB. Currently, there is not a dedicated maintainer or group of maintainers developing solutions for automation. Efforts have been split between the old and new engines. Automation is extremely important and large enough to deserve special attention, rather than being buried in OHC.

There are obviously many more things to discuss, design, develop, etc., but everything is dependent on whether or not automation stays in OHC. It would be helpful to get this issue sorted soon, so that users can begin their migrations. When word of the removal of Joda spreads, it will cause many to begin revamping their rules, which will hopefully incite them to migrate to Jython... especially if there is a clear roadmap for automation.

**Comment by openhab-5iver** [comment]
@kaikreuzer, the indecision here is holding me back from developing! Or am I still arguing my case and the decision is already final? I'm surprised by the lack of input from the other @openhab/core-maintainers.I see this issue as a cornerstone that needs to get placed before moving forward. In looking back, I could have gotten a lot done in the last month and a half!

Even though I don't like the idea of putting this into addons, I'm excited at the possibility of getting it out in the next 2.5.x release. The manual installs in the forum have gone very well.

**Comment by J-N-K** [comment]
I would like to bring up another point which I think supports @openhab-5iver. Currently we need to add jython-stand-alone-2.7.0.jar to the bootclasspath to use jython. This completely breaks scripting in java11. I didn‘t find an easy way to work around this, but there is a slim machen dependency for jython. Having scripting separated into different addons, we could probably use that one and leave the decision on which engine shall be used to the user. Maybe defaulting to jython.

**Comment by kaikreuzer** [comment]
Sorry to be so slow in responding @openhab-5iver, there are far too many things on my plate in the past weeks, I cannot follow up as much as I'd like to.

I still do not really get your point here - or maybe let me say that I rather see many issues arising if we completely follow your suggestion.
- The NGRE is foremost just a pluggable framework for assembling rules (as such, it is not at all about scripting). Moving this out of core sounds a bit like moving the Thing/Binding API out of core, because it is only relevant for bindings, which could be considered optional with the same argument.
- Also note that the model.script and model.rule bundles are terribly complex to build and maintain as it requires Xtext code generation and a lot of effort to adapt with every upgrade. Effectively, my impression is that only @wborn is capable of handling changes and upgrades on this these days. Adding that complexity to yet another repo sounds very undesirable.
- Yes, script engine factories, module types, templates, etc. are fine for me to be maintained somewhere else. But here again, we should decide, what we consider "core", i.e. what the users can expect to always find and what is purely optional, because just relevant for a few users in special cases, which would make them "add-ons" since we must provide a way for the users to browse and install them.

> Currently we need to add jython-stand-alone-2.7.0.jar to the bootclasspath to use jython. This completely breaks scripting in java11.

This sounds rather like an argument that we should stick to DSLs as our preferred format and move Jython support to some other repo, so that it becomes purely optional, since it seems to be fragile anyhow and we never know whether a future combination of Jython+Java will work at all. If we would not also have the discussion that Jython should fully replace the DSLs, I would agree with you that this issue supports @openhab-5iver case here.

So in summary, I more or less just said the same as in https://github.com/openhab/openhab-core/issues/1319#issuecomment-573217169.
The only change I can live with is to move the script support (i.e. the engine implementations and all their libs) to some other place (yes, maybe a separate repo, if nothing else depends on them).

Reading your original argument of what makes you want a separate repo, this is all about the concrete script engine implementations and their ecosystem - so would you agree that this could be a compromise?

@wborn & @cweitkamp I'd also very much interested in your opinions here at it is an important decision for the overall openHAB distro structure.

**Comment by wborn** [comment]
I agree with @kaikreuzer that it's best to keep the automation APIs and rule engine in openhab-core. Automation is what this software is all about and if automation was something optional there wouldn't be an A in openH**A**B.

It makes a lot of sense for OH3 to standardize on Jython as preferred NGRE language because it is already used by many, has a good helper library and there are many examples.

I think we should always keep the possibility open to also use other JSR223 languages with openHAB. It is a unique selling point that openHAB users can program in their preferred language of choice instead of being forced a language upon them. Existing users (me included) may also have already invested their time in one of those other languages.

When openhab-core provides all the necessary automation APIs each JSR223 language add-on is optionally installable so it would make sense to add these to openhab-addons for now. We can always split these off to another repo if we have reasons for this.

If we want OH3 to more tightly integrate with JSR223 languages we probably also want to make the helper libraries official and move the [openhab-helper-libraries](https://github.com/openhab-scripters/openhab-helper-libraries) into the openhab organization. I would initially keep it a separate repo with its own maintainers until there is a clear bigger picture of how this all integrates. If that repo publishes versioned artifacts we can use these as dependencies in JSR223 add-ons or the distro and later on add them to the CI pipelines.

**Comment by J-N-K** [comment]
Regarding jython: good luck. If you use google with „jython osgi“ you easily see that this is very difficult. With the removed/changed boot classpath in Java11 the way we integrated it before is not working anymore. I spend some hours trying to figure out how to do that but failed. OSGIifing jython is also difficult because it is a multi-version jar which is not supported by bnd.

I changed all rules to JavaScript.

**Comment by jpg0** [comment]
I want to say that I disagree that OH3 "_should standardize on Jython as preferred NGRE language_". I have a few reasons for this:
* Picking **one** language often results in making things unworkable for new languages. Having two (at least) options will mitigate much of this. This does not necessarily result in greater complexity (for example I see decisions with the current directory layout which will make things tough for JS).
* Home Assistant is already pure Python. I'm not much of a fan of Python, which is why I originally chose OH over HA. If I was a fan of Python, I would expect that I would have just gone straight to HA. Supporting multiple languages for rules I regard as a huge plus for OH.
* The current future of scripting on the JVM is Graal. This already provides a rich abstraction that OH can use. Whilst GraalJS is mature already, GraalPython is not yet there (although may be for what OH needs, it's not clear), although I'd be pretty surprised if it did not replace Jython. (I may be able to create a bundle with GraalPython support if it's useful.)

**Comment by J-N-K** [comment]
Nothing to add from my side to what @jpg0 said.

**Comment by openhab-5iver** [comment]
> * The NGRE is foremost just a pluggable framework for assembling rules (as such, it is not at all about scripting). Moving this out of core sounds a bit like moving the Thing/Binding API out of core, because it is only relevant for bindings, which could be considered optional with the same argument.

I still don't see this the way you do. A pluggable framework can still be completely optional. If it is optional, then IMO it is not suitable for OHC. With the right automation bundle(s), you would not need _any_ rule engine at all. Rather than the Thing/Binding API, I think the UIs are a better comparison. IMO, the core framework should exist to provide modularity for optional components like the UIs and automation. If automation is to be kept in OHC, then why couldn’t the same argument be made for keeping the UIs there too?


> * Also note that the model.script and model.rule bundles are terribly complex to build and maintain as it requires Xtext code generation and a lot of effort to adapt with every upgrade. Effectively, my impression is that only @wborn is capable of handling changes and upgrades on this these days. Adding that complexity to yet another repo sounds very undesirable.

Based on several discussions in GH and posts in the forum, I had the impression that Xtext and the old rule engine would be removed in OH 3.0. This was an impetus for my requesting that automation be removed from OHC. If our user base and development efforts are to be split between the old and new rule engines, then both will be less than what they could be. I do not fully comprehend the increased complexity that you refer to; but if this exists, then it bolsters the case for their removal and fits well into the theme of consolidation for OH 3.0.


> * Yes, script engine factories, module types, templates, etc. are fine for me to be maintained somewhere else. But here again, we should decide, what we consider "core", i.e. what the users can expect to always find and what is purely optional, because just relevant for a few users in special cases, which would make them "add-ons" since we must provide a way for the users to browse and install them.

It sounds as though me, you and Wouter are in agreement with removing automation from OHC, with the exception of the packages containing the rule engines themselves. It would be nice to know the thoughts of @cweitkamp, but at least this is a start! Now we just need to decide where these bundles will reside. 

Where do you think Scripted Automation falls in this? Should it be left with the rule engine or moved with the rest? My preference is to move everything possible out of OHC and into a new repo (openhab-automation).


>> Even though I don't like the idea of putting this into addons, I'm excited at the possibility of getting it out in the next 2.5.x release.

> Err, here you must have misunderstood something. Everything we do here is on master, i.e. it is relevant for 3.x, but definitely not for 2.5.x.

My comment was in expectation of this issue being rejected and that the JythonScriptEngineFactory would need to be moved from OHC to openhab-addons. This would not require any changes to OHC, so could be included in the next 2.5.x release. If this is the direction we’re going to go, this will be a quick update. User testing has gone very well and it will go even better when it can be installed through Paper UI. 

> Reading your original argument of what makes you want a separate repo, this is all about the concrete script engine implementations and their ecosystem - so would you agree that this could be a compromise?

My apologies if I was not clear, but my proposal is to remove _all_ of automation from OHC, not just Scripted Automation and/or the ScriptEngineFactories. At this point, I would agree to anything in order to get development moving forward, but my opinion that automation is completely optional and should not be located in OHC has not changed. I have voiced my opposing viewpoint and will move forward.


> Automation is what this software is all about and if automation was something optional there wouldn't be an A in openHAB.

At first glance, this seems logical. But while the **A** in openHAB does stand for **A**utomation, you can’t separate it from the **B**us. The rule engines rely on the **H**ome **A**utomation **B**us provided by OHC, and _that_ is what is not optional. The decision should be left to the user as to which optional component (rule engine, script engine, UI, addon, etc.) they attach to their **H**ome **A**utomation **B**us.

I know… this is arguing semantics a bit… but it’s also a simple illustration of what I understand the OHC to be and why I think of automation as an optional component.


> Regarding jython: good luck.

I can confirm that a manual installation (adding the Jython jar into the bootclasspath) of either Jython 2.7.2b3 or 2.7.1 does not work in OH 3.0 with JDK11. I suspect it is not an issue in Jython but with OH. Jython 2.7.2b3 works fine with OH 2.5.x with JDK8. However, the JythonScriptEngineFactory bundle using Jython 2.7.2b3 works fine with OH 3.0 with JDK11. I will push an update to the bundle that bumps Jython to 2.7.2b3 and removes the helper libraries when things here have been settled.


> I want to say that I disagree that OH3 "should standardize on Jython as preferred NGRE language".

I appreciate your comment, but let’s take this up somewhere else, as this issue is not the place for such a discussion. Keep in mind though, [that ship sailed over a year ago](https://community.openhab.org/t/directions-for-openhab-3/67372)… along with thousands of man-hours of effort building up the Jython helper libraries, documentation, and migrating/supporting users.

**Comment by kaikreuzer** [comment]
>  I had the impression that Xtext and the old rule engine would be removed in OH 3.0

Xtext will stay, the old rule engine will be removed.

> removing automation from OHC, with the exception of the packages containing the rule engines themselves.

Which effectively means that most current automation bundles will stay in core and mainly the jsr223 scripting features move to somewhere else.

> Where do you think Scripted Automation falls in this?

As mentioned before: The APIs should stay in core, the concrete implementations should be treated as add-ons.

>If we want OH3 to more tightly integrate with JSR223 languages we probably also want to make the helper libraries official and move the openhab-helper-libraries into the openhab organization. I would initially keep it a separate repo with its own maintainers until there is a clear bigger picture of how this all integrates. If that repo publishes versioned artifacts we can use these as dependencies in JSR223 add-ons or the distro and later on add them to the CI pipelines.

I very much like this suggestion of @wborn and think that should be a good way to move forward with the existing helper libraries.

>>  I disagree that OH3 "should standardize on Jython as preferred NGRE language"
> I appreciate your comment, but let’s take this up somewhere else, as this issue is not the place for such a discussion. Keep in mind though, that ship sailed over a year ago… 

I don't think that the reference you provide here contradicts @jpg0's statement - I think he gave very strong arguments in https://github.com/openhab/openhab-core/issues/1319#issuecomment-599159277, which I consider as very valid points and I personally wouldn't feel good betting openHAB's future on the smooth integration of Python and that all its community is embracing Python as a language. In above link I said "Jython-rules should be made a first class citizen", which means that all the work that went into this is highly appreciated and that there are obviously many supporters that prefer writing their rules in Python instead of the Rule DSL - so this option should become a fully supported. It isn't saying that it has to be the preferred option, though. The same is true for Javascript: I welcome all the work that @jpg0 is doing on this with the GraalVM as the new Java-way of scripting and there's likewise a huge bunch of people that prefer writing their rules that way. 
I further said: "The NG rule engine should be the only one - existing Xbase-rules should be supported by it". I have just started a PR to remove the old rule engine and it seems it won't be too hard to make Xbase-rules available in the NGRE, so I think this will be highly appreciated by the community.

But back to the main question we have to solve: Where and how should those scripting implementations be moved? I see that you (both) have meanwhile created PRs for openhab-addons, which is nice.
What we are still lacking, though is an agreement on the categories of add-on that we want to have/introduce for automation - those categories will have to be supported by the UI for installation, by the future marketplace, etc., so we should spend some thoughts on this topic.
So far, [this](https://github.com/openhab/openhab-core/blob/3811cb95604f69f6a3d77da2caf391d81b05458e/bundles/org.openhab.core.karaf/src/main/java/org/openhab/core/karaf/internal/KarafExtensionService.java#L57-L64) is our list of categories - the only one related to automation are the "actions", which is now deprecated in openHAB 3, since we moved those into the bindings. We are thus completely blank right now.

One option would be to add an "automation" category, where we would put any kind of automation-related add-on into. Those could be script-language support add-ons (like for Python and JS), optional module types as well as rule templates. My feeling is that at least rule templates should be a separate category, because there is a chance for a very lively community around it, which produces many many such add-ons (for the Eclipse Marketplace we actually already had "rule template" as a category).

What do others think? For defining the categories, I would love to have a common decision of @openhab/core-maintainers, @openhab/add-ons-maintainers and @openhab/webui-maintainers as it will have impact on all of them.

**Comment by splatch** [comment]
> > I had the impression that Xtext and the old rule engine would be removed in OH 3.0
> 
> Xtext will stay, the old rule engine will be removed.

While one, another or both might stay it shouldn't be made a dependency in the core of project. **We need to answer question if openhab-core ships reusable framework, or a product itself**. From what I remember in the "Road ahead ESH" statements, project core should remain a framework. So a proper place to mark rule engine of any kind as a shipped feature is distribution build and not core. If we make clear that, then it doesn't matter where from automation part comes from, is it coming from core or another repository. For end users it is important that it comes in as a part of their distribution.

My small suggestion here is following - if `openhab-core-base` feature is not intended for re-use then it should be removed. If its intended then it should be composed from smaller features and not bundle listing.
https://github.com/openhab/openhab-core/blob/2.5.0/features/karaf/openhab-core/src/main/feature/feature.xml#L28

Cheers,
Łukasz

**Comment by kaikreuzer** [comment]
> We need to answer question if openhab-core ships reusable framework, or a product itself.

Nothing has changed in this regard. openhab-core is a set of bundles that can be picked and used in other products than openHAB as well. To be flexible, it makes a lot of sense for such products to define their own features (if they want to use Karaf at all, which isn't necessary or even recommended either).

**Comment by kaikreuzer** [comment]
> What do others think? For defining the categories, I would love to have a common decision of @openhab/core-maintainers, @openhab/add-ons-maintainers and @openhab/webui-maintainers as it will have impact on all of them.

@wborn, @ghys, @cweitkamp and @openhab/add-ons-maintainers, I'd love to hear your position on suitable add-on categories as per https://github.com/openhab/openhab-core/issues/1319#issuecomment-607474276!

**Comment by ghys** [comment]
> One option would be to add an "automation" category, where we would put any kind of automation-related add-on into. Those could be script-language support add-ons (like for Python and JS), optional module types as well as rule templates. My feeling is that at least rule templates should be a separate category, because there is a chance for a very lively community around it, which produces many many such add-ons (for the Eclipse Marketplace we actually already had "rule template" as a category).

Having an "automation" category for script engines looks fine to me.
For templates, they're basically JSON files, not .jars, right? If so, I'd like to consider defining a general-purpose "templates" category not only focus on rule templates, but other types of templates or blueprints for other types of objects as well. The new pages and widgets in the UI come to mind.
How would these be sourced anyway, would they be bundled in the distribution as well, or would there be another provider (like the marketplace) to publish and retrieve them?

**Comment by kaikreuzer** [comment]
> For templates, they're basically JSON files, not .jars, right?

It depends. Such templates could potentially come with their own module types, which could include complex Java logic and would be bundled together with them. Alternatively, we could say that module types can be provided as "automation" add-ons and we would have to define a dependency mechanism from templates onto module types (would help to reuse them and avoid name clashes, but obviously makes stuff more complex).

I don't think we should mix templates of different domains (automation/ui/...), though. Our add-on categories also refer to code skeletons, how-to tutorials, etc and a rule template has simply nothing in common with an ui template besides the name.

But I think you are raising a good point: Templates might not be a suitable add-on category at all, they probably qualify for something on their own - we won't package them as OSGi bundles, but rather have JSON files with maybe some additional metadata. They might need their own UI browsing and installing mechanism, similar to what you did on HABPanel already.

**Comment by wborn** [comment]
The scripting engines could become their own category if one day there are many. But if there are only a few they can part of misc.

It would help a great deal if we can introduce or restructure categories at a later time. So we should try to facilitate that e.g. by making it possible to update Karaf feature names.

For new users each category type probably doesn't say a lot. So if there are many they will have a hard time finding add-ons. For them it would probably be easier if they can just search over all add-ons and install it similar to App Stores or the Add-ons/Extensions for browsers.

**Comment by kaikreuzer** [comment]
I didn't think of that, but you are right: If we are mainly talking about Jython+JS as scripting support add-ons, we could just put them under "misc". It does not answer the question about the Java package name / bundle name, though as most "misc" add-ons are actually "org.openhab.io" (with a few exceptions). What I would like to get to is a naming convention like `org.openhab.<addontype>.<id>` and I think we are pretty close to this in openhab-addons (and we were in the past even closer, when we still had sub-folders, which had the name of the add-on type).
That's why I would tend to go for `org.openhab.automation.jythonscripting` for the script engine factories and treat the "automation" category not too strict wrt what kind of automation feature is provided by the add-on.

> It would help a great deal if we can introduce or restructure categories at a later time. So we should try to facilitate that e.g. by making it possible to update Karaf feature names.

If you think of a category merely as a marker, then this should be easy by changing the Karaf feature name. But I was actually also more thinking of categories wrt add-on skeleton generation and developer documentation - those "add-on categories" would be rather static and now easy to be changed. And imho, we should try to stick to a few such categories, so that we do not end up with just "OSGi bundles" that do whatever they want...

> For them it would probably be easier if they can just search

Sure, but I think this is easy to do in the UIs and it is not what I was after in this discussion.

**Comment by wborn** [comment]
I'm fine with adding `automation` to these packages and  bundle IDs. It relates to what we have in the core and it is more generic than using something with `script(ing)(engine)`

**Comment by kaikreuzer** [comment]
Ok, then unless anybody vetos in the next 48h, let's go with a new "automation" category and use
`org.openhab.automation.jythonscripting` for https://github.com/openhab/openhab-addons/pull/7208 and
`org.openhab.automation.jsscripting` for https://github.com/openhab/openhab-addons/pull/6399.

@openhab-5iver and @jpg0, would you been ok with that choice?

**Comment by openhab-5iver** [comment]
I agree with the "automation" category, but I think we will have a big messy list of addons if we do not plan a little more for the other addons that will go under `org.openhab.automation`.

- ScriptEngineFactories
- core helper libraries
- community helper libraries
- automation bundles
- ModuleTypes

Is there a reason why we coulnd't breaking things down a couple more levels?

```
# There are others, but I think these will be the most commonly used.
org.openhab.automation.scriptenginefactory.jython
org.openhab.automation.scriptenginefactory.nashorn
org.openhab.automation.scriptenginefactory.graaljs
org.openhab.automation.scriptenginefactory.groovy
org.openhab.automation.scriptenginefactory.kotlin
org.openhab.automation.scriptenginefactory.jruby

# These are for the core packages and all of them will be replaced by the scripting API.
org.openhab.automation.helperlibraries.jython
org.openhab.automation.helperlibraries.nashorn
org.openhab.automation.helperlibraries.graaljs
org.openhab.automation.helperlibraries.groovy
org.openhab.automation.helperlibraries.kotlin
org.openhab.automation.helperlibraries.jruby

# These are some example Jython community packages, of which there are currently 11 and several more pending approval.
# These will be the most plentiful and there will potentially be community packages for each language, but none have been submitted yet.
org.openhab.automation.helperlibraries.jython.areatriggersandactions
org.openhab.automation.helperlibraries.jython.idealarm

# These are Java-based automation packages and TMK none exist yet.
org.openhab.automation.exampleautomationbundle

# I question ModuleType here though, since these would be better off with the others in the rule engine... or those could be moved over here.
org.openhab.automation.moduletype.<id>
```
Rule templates would still need to be somewhere in a repository and maintained, so a `/ruletemplates/<name>` directory off the root should work.

**Comment by kaikreuzer** [comment]
>  the other addons that will go under org.openhab.automation.

Per discussion above, there shouldn't be only two, which would be ok to gather in one "automation" category:

- ScriptEngineFactories + core helper libraries (as one add-on)
- ModuleTypes

"community helper libraries" rather sound like something for a future marketplace and yes - for the marketplace, we might maybe have additional types, such as rule templates or ui templates or widgets as discussed above.

For the official core, I think that a proliferation of new add-on categories or even hierarchies will cause confusion and won't help us any further, while complicating the add-ons concept immensely (think of all the naming conventions for features, the automatic doc/website build, etc.).

**Comment by openhab-5iver** [comment]
The ScriptEngineFactories, the core libraries, and the community libraries need to go into their own individual addons for the reasons that I provided in the [Jython Helper LIbrary addon PR](https://github.com/openhab/openhab-addons/pull/7210#issuecomment-617719792).

**Comment by kaikreuzer** [comment]
> The ScriptEngineFactories, the core libraries, and the community libraries need to go into their own individual addons

What you are saying is that there is no dependency between them? Because this is a hard requirement for any add-on.
I honestly do not understand, what help a core library would be without having the general jythonscripting feature in place...

**Comment by openhab-5iver** [comment]
I view J/Python as an external dependency to the helper libraries. A helper library add-on needs Jython, just like the JDBC persistence add-on needs some flavor of SQL database. The helper libraries just need Jython in the classpath, which could be in the form of a full or standalone manual install, or through a Jython 2.7.2, 3.x, or graal-python add-on. However J/Python is installed, the helper libraries will function. If the libraries were locked to a specific version, then we would need separate add-ons for every flavor of J/Python.

**Comment by kaikreuzer** [comment]
Sorry if I may ask stupid questions, but it still isn't at all clear to me what all of this is about.
To me, the discussion here was around the new rule engine and how to allow users to use Python as a scripting language to define rules.
My assumption was that we can offer a `org.openhab.automation.jythonscripting` add-on, which brings this capability to openHAB.
Do I understand your last comment correctly that this isn't really the case, but that J/Python still needs to be provided "some way" by the user?
And do I get you right that you consider "helper libraries" to be some independent add-on that is not related to the jytonscripting add-on? If so, what would be the use-case of it?

**Comment by openhab-5iver** [comment]
> Sorry if I may ask stupid questions, but it still isn't at all clear to me what all of this is about.
To me, the discussion here was around the new rule engine and how to allow users to use Python as a scripting language to define rules.
My assumption was that we can offer a org.openhab.automation.jythonscripting add-on, which brings this capability to openHAB.

This discussion has gotten WAY off topic from what I had originally presented. My intention was to discuss creating a repository that would hold everything related to automation, where the maintainer(s) could focus specifically on automation. I feel this is very much needed for openHAB to evolve. You and @wborn preferred to keep the rule engines in OHC, so the scope of the discussion turned to the creation of a repository for automation, but with the exception of the current rule engine. An openHAB repository has not yet been created for this purpose. 

Since then, discussion has been caught up in the details of the namespace to use, scripted automation, and my PRs. These were still important discussions to have, at least for me. However, we are far from the topic that I had wanted to resolve before I began to contribute to automation in OH 3.0. Seven months later, we have perfectly illustrated why OH automation is not progressing like it could be and the benefits of having its own repository with maintainers, like me, that could be focused on automation.

> Do I understand your last comment correctly that this isn't really the case, but that J/Python still needs to be provided "some way" by the user?
And do I get you right that you consider "helper libraries" to be some independent add-on that is not related to the jytonscripting add-on? If so, what would be the use-case of it?

No, the Jython add-on will setup Jython for you. However, there are many other flavors and installation methods available. Yes, HLs should be a separate add-on for several reasons, but mainly so that people would still be able to use them if they choose not to use the Jython add-on. The Jython HL add-on has no dependency on _the Jython add-on_, but it does need Jython to be setup in some way in order to use them.

Before 2.5.0 was released, I had submitted a PR that combined Jython and the core and community HLs, but it was best to separate them. I currently have two PRs for addons, where one [adds Jython 2.7.2 standalone](https://github.com/openhab/openhab-addons/pull/7208) and another that [adds the Jython helper libraries](https://github.com/openhab/openhab-addons/pull/7210), which will be replaced by the scripting API. I have many more automation addons to submit.

Moving forward, I suggest that we have those off-topic discussions (namespaces, contents of HL add-ons, etc.) in my PRs and finalize here the topic of a separate repository for automation.

**Comment by DerOetzi** [comment]
I didn't noticed this RFC before. Just some thoughts about jython support as "default" rules language. Although I switched all of my rules to jython some months ago I'm not really convinced about the future of jython for following reasons:

- If you look at the release speed of jython, this has really slowed down since 2017. For me it looks a little bit like a dead project
- Version 2.7.0 has some major issues like pip not working
- Version >=2.7.1 have a huge performance issue which wasn't fixed since 2017 https://bugs.jython.org/issue2604
- Most important point: officially python 2.7 had an announced end of life date 2015 which was moved to 1st January 2020 by https://legacy.python.org/dev/peps/pep-0373/ which although is in the past now and the 3.X compatibly jython development has stucked since years. 
- Actual Pip is throwing a deprecated warning announcing end of 2.7 support for the beginning of next year

So I really like the possibility to write rules in python and especially the great helper libraries but I'm pretty sure jython does not have a future from my point of view.
